### PR TITLE
Reduce number of typeEquiv checks

### PR DESCRIPTION
**Pull request by forki**

### CODE REVIEW

**Comment by En3Tho** [comment]
Do we have some sort of benchmark suite to see if these changes are helpful? I'm planning to try finding opportunities for optimizations too and I wonder if I find one, I can be assured stuff I code really helps

**Comment by cartermp** [comment]
The closes thing we have is CompilerServiceBenchmarks: https://github.com/dotnet/fsharp/blob/main/tests/benchmarks/CompilerServiceBenchmarks/Program.fs

We use benchmarkdotnet to experiment against a few different scenarios. It may need to be updated again now that some files have been rearranged.

If the benchmark doesn't yield anything meaningful then that's fine though. Sometimes a single change is marginal but when you combine it with 3-4 other marginal changes it's no longer marginal. So long as it's obvious that there is less work being done in the code then we'd be inclined to take it.

That said, the more evidence, the better!

**Comment by jonsequitur** [comment]
We've started on building up a benchmark suite here: https://github.com/dotnet/performance/pull/1546.

Some discussion might be helpful here to make sure that we're designing benchmarks that will be useful for historical comparison over time as code is refactored.

**Comment by smoothdeveloper** [comment]
@jonsequitur I've reported a non-fruitive exploration in #8260 about experience of compiling FSharpPlus library unit tests.

The unit test codebase is arguably not representative of general F# codebase but it provides a very good test ground for overload resolution complexity and would very likely highlight performance regression in that logic.

I don't know for sure if the F# compiler team sees that as salliant but it felt to me we should take a snapshot of that codebase and use it as a benchmark.

If you feel it may have relevance for dotnet/performance#1546 I'll pay a closer look at that repository.

File: src/fsharp/CheckDeclarations.fs [hunk 1]
```diff
@@ -2374,12 +2374,18 @@ let TcMutRecDefns_Phase2 (cenv: cenv) envInitial bindsm scopem mutRecNSInfo (env
                   
                   if not (tcref.HasInterface g ity') then 
                       error(Error(FSComp.SR.tcAllImplementedInterfacesShouldBeDeclared(), ity.Range))
-                  if (typeEquiv g ity' g.mk_IComparable_ty && Option.isSome tcref.GeneratedCompareToValues) || 
-                      (typeEquiv g ity' g.mk_IStructuralComparable_ty && Option.isSome tcref.GeneratedCompareToWithComparerValues) ||
-                      (typeEquiv g ity' ((mkAppTy g.system_GenericIComparable_tcref [ty])) && Option.isSome tcref.GeneratedCompareToValues) ||
-                      (typeEquiv g ity' ((mkAppTy g.system_GenericIEquatable_tcref [ty])) && Option.isSome tcref.GeneratedHashAndEqualsWithComparerValues) ||
-                      (typeEquiv g ity' g.mk_IStructuralEquatable_ty && Option.isSome tcref.GeneratedHashAndEqualsWithComparerValues) then
+                   
+                  let generatedCompareToValues = Option.isSome tcref.GeneratedCompareToValues
```
Comment (dsyme) [inline src/fsharp/CheckDeclarations.fs:2378]: `tcref.GeneratedCompareToValues.IsSome` in each cast please.

Comment (forki) [reply src/fsharp/CheckDeclarations.fs]: done

**Review by cartermp — APPROVED** [review]

**Review by KevinRansom — APPROVED** [review]
Thanks for this.

**Pull request merged**